topics/cluster-tutorial.md

```diff
@@ -928,16 +928,53 @@ it with an updated version of Valkey. If there are clients scaling reads using
 replica nodes, they should be able to reconnect to a different replica if a given
 one is not available.
 
-Upgrading primaries is a bit more complex, and the suggested procedure is:
-
-1. Use `CLUSTER FAILOVER` to trigger a manual failover of the primary to one of its replicas.
-   (See the [Manual failover](#manual-failover) in this topic.)
-2. Wait for the primary to turn into a replica.
-3. Finally upgrade the node as you do for replicas.
-4. If you want the primary to be the node you just upgraded, trigger a new manual failover in order to turn back the upgraded node into a primary.
-
-Following this procedure you should upgrade one node after the other until
-all the nodes are upgraded.
+Upgrading primaries is a bit more complex. The suggested procedure is to trigger
+a manual failover to turn the old primary into a replica and then upgrading it.
+
+A complete rolling upgrade of all nodes in a cluster can be performed by
+repeating the following procedure for each shard (a primary and its replicas):
+
+1. Add one or more upgraded nodes as new replicas to the primary. This step is
+   optional but it ensures that the number of replicas is not compromised during
+   the rolling upgrade. An alternative is to upgrade one replica at a time and
+   have fewer replicas online during the upgrade. To add a new node, use
+   [`CLUSTER MEET`](../commands/cluster-meet.md) and [`CLUSTER
+   REPLICATE`](../commands/cluster-replicate.md) or use `valkey-cli` as
+   described under [Add a new node as a replica](#add-a-new-node-as-a-replica).
+
+2. Upgrade the existing replicas you want to keep by restarting them with the
+   updated version of Valkey. If you're replacing all the old nodes with new
+   nodes, you can skip this step.
+
+3. Select one of the upgraded replicas to be the new primary. Wait until this
+   replica has caught up the replication offset with the primary. You can use
+   [`INFO REPLICATION`](../commands/info.md) and check for the line
+   `master_link_status:up` to be present. This indicates that the sync with the
+   primary is complete.
+
+4. Check that the new replica is known by all nodes in the cluster, or at least
+   by the primaries in the cluster. You can send [`CLUSTER
+   NODES`](../commands/cluster-nodes.md) to each of the nodes in the cluster and
+   check that they all are aware of the new node. Wait for some time and repeat
+   the check if necessary.
```

Comment:
Might be handy to to build it in as a utility in the `valkey-cli` to perform this check.

Reply:
Yes, that'd be good, as a reference implementation. The first step is docs though IMHO.

Even better would be to have an official open source kubernetes operator to orchestrate things like rolling upgrade and scaling, even autoscaling, by deploying new pods as necessary.

Reply:
Definitely.